### PR TITLE
[TypeChecker] SE-0352: Require coercion if result type contains existential(s) that would loose generic requirements

### PR DESCRIPTION
**Pull request by xedin**
Implements SE-0352 revision - require explicit `as` coercion when existential
erasure loses information. 

For example:

```swift
protocol P {
  associatedtype A
}

protocol Q {
  associatedtype B: P where B.A == Int
}

func getB<T: Q>(_: T) -> T.B { ... }

func test(v: any Q) {
  let _ = getB(v) // <- produces `any P` which loses A == Int
}
```

### CODE REVIEW

**Comment by xedin** [comment]
@DougGregor Please take a look to make sure that this is going in the right direction, I still need to implement this diagnostic for Self erasure cases but it shouldn't be too bad.

File: lib/Sema/CSFix.cpp [hunk 1]
```diff
@@ -2151,3 +2151,127 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
...
+          RequiresCoercion |= hasConstrainedAssociatedTypes(opened->second);
+          return RequiresCoercion ? Action::Stop : Action::SkipChildren;
+        }
+      }
+
+      return Action::Continue;
+    }
+
+  private:
+    bool hasConstrainedAssociatedTypes(ArchetypeType *archetypeTy) {
```
Comment (xedin) [inline lib/Sema/CSFix.cpp:2244]: I realized that this condition is incorrect, I need to check whether any member in the chain of associated types described by dependent member ends up having a where clause.

Comment (xedin) [reply lib/Sema/CSFix.cpp:2257]: Fixed!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSFix.cpp [hunk 2]
```diff
@@ -2151,3 +2152,162 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
...
+
+      return Action::Continue;
+    }
+
+  private:
+    bool hasConstrainedAssociatedTypes(DependentMemberType *member) {
+      auto *assocType = member->getAssocType();
+
+      // If this member has associated type requirements, we are done.
+      if (assocType->getTrailingWhereClause())
```
Comment (DougGregor) [inline lib/Sema/CSFix.cpp:2231]: This is looking at the syntactic construct (the trailing where clause), which is not serialized and may not have complete information. Can we instead look at the generic signature to see if there are any requirements that mention this dependent member type?

File: lib/Sema/CSFix.cpp [hunk 3]
```diff
@@ -2151,3 +2152,162 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
...
+        return getBaseTypeOfDependentMemberChain(DMT);
+
+      return base;
+    }
+
+    static bool hasConstrainedAssociatedTypes(ExistentialLayout layout) {
+      for (auto *protocol : layout.getProtocols()) {
+        if (llvm::any_of(protocol->getAssociatedTypeMembers(),
+                         [](const auto *assocTypeDecl) {
+                           return bool(assocTypeDecl->getTrailingWhereClause());
```
Comment (DougGregor) [inline lib/Sema/CSFix.cpp:2256]: Same concern as above: `getTrailingWhereClause` is more syntactic than semantic.

File: lib/Sema/CSDiagnostics.cpp [hunk 4]
```diff
@@ -8143,3 +8143,24 @@ bool DefaultExprTypeMismatch::diagnoseAsError() {
...
+      diag::candidate_result_requires_explicit_coercion, ErasedResultType);
+  fixIt(diagnostic);
+  return true;
+}
+
+void MissingExplicitExistentialCoercion::fixIt(
+    InFlightDiagnostic &diagnostic) const {
+  auto printOpts = PrintOptions::forDiagnosticArguments();
+  diagnostic.fixItInsertAfter(getSourceRange().End,
+                              "as " + ErasedResultType->getString(printOpts));
```
Comment (DougGregor) [inline lib/Sema/CSDiagnostics.cpp:8165]: Could we insert parentheses in cases where the expression being coerced is an argument to a function that can be opened?

File: lib/Sema/CSFix.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSFix.cpp:2231]: Ah, I didn't know that this was the case, will fix it before merging this.

File: lib/Sema/CSDiagnostics.cpp (hunk 4 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp:8165]: Sure!

**Review by DougGregor — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/CSDiagnostics.cpp (hunk 4 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Implemented!

**Comment by xedin** [comment]
@swift-ci please test

**Review by DougGregor — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Pull request merged**